### ISSUE TITLE
`Error: Not Found`

### ISSUE THREAD

**Issue by polarathene** [micalevisk/last-issue-action#5]
I am not sure why, but the action is failing to find an issue when it exists with the label. Taking the `GET` request from the error log below and viewing it in the browser shows the issue listed.

The repo is private, but AFAIK that should not be an issue? Inspecting the action log the `GITHUB_TOKEN` permissions looks fine.

---

**Error:**

```
GET /repos/polarathene/test-issue-creation/issues?state=open&labels=automated-issue&sort=updated&direction=desc&per_page=1&limit=1 - 404 in 74ms
Error: Not Found
Error: Something went wrong!
```

**`.github/workflows/test-issue-search.yml`:**

```yml
name: 'Test - Issue Search'
on: workflow_dispatch

jobs:
  test-issue-search:
    runs-on: ubuntu-22.04
    steps:
      - name: Find the last open issue with label automated-issue
        uses: micalevisk/last-issue-action@v1.2
        with:
          state: open
          labels: automated-issue
```

---

## Resolved

Although 403 is a more common error status to encounter when a Github Token is lacking permissions, the REST API used here [may return 404 as a security practice apparently](https://stackoverflow.com/questions/65280412/octokit-put-repos-owner-repo-contents-path-status-404/65285503#65285503):

> _A 404 generally means an authentication issue: for security reason, the API pretends there is no such resource (even if it exists) when it detects you are not allowed to access/update it._

**Comment by polarathene** [comment]
https://github.com/micalevisk/last-issue-action/blob/a779b00cf9770e42a5ff23207d40c68a3f02a94f/README.md?plain=1#L46-L48

Despite what the README says here, this was required. After adding the ENV, the action successfully got a `200` response.

I've not tested the action on a public repo yet, so it may be due to using a private repo, or some other reason that secret was not available to the action via ENV.

---

This was not required for `peter-evans/create-issue-from-file`, it manages to use the token automatically:

- [Assigning the default to `input.token`](https://github.com/peter-evans/create-issue-from-file/blob/b910fa165602be1908fb8b1cd372f1d7109206f4/action.yml#L3-L6) (uses `${{ github.token }}`)
- [Reading `input.token`](https://github.com/peter-evans/create-issue-from-file/blob/b910fa165602be1908fb8b1cd372f1d7109206f4/src/main.ts#L11).

Since they handle `input.repository` in exactly the same way, and you [likewise retrieve this information like you do with the `GITHUB_TOKEN` via ENV](https://github.com/micalevisk/last-issue-action/blob/a779b00cf9770e42a5ff23207d40c68a3f02a94f/src/index.ts#L11-L14), I'm guessing the token is not available via ENV due to being considered a secret, but valid when defined in `action.yml` like `peter-evans/create-issue-from-file` has done?

**Comment by micalevisk** [comment]
I guess that env is required as I don't think we can use `${{ github.token }}` as default for env. vars. in the `action.yml` (https://docs.github.com/actions/creating-actions/metadata-syntax-for-github-actions) file :/ 

I'll edit the README then

<blockquote><img src="https://github.githubassets.com/images/modules/open_graph/github-logo.png" width="48" align="right"><div><img src="/assets/cb-803/images/site/favicon.svg" height="14"> GitHub Docs</div><div><strong><a href="https://ghdocs-prod.azurewebsites.net/en/actions/creating-actions/metadata-syntax-for-github-actions">Metadata syntax for GitHub Actions - GitHub Docs</a></strong></div><div>You can create actions to perform tasks in your repository. Actions require a metadata file that uses YAML syntax.</div></blockquote>

**Comment by polarathene** [comment]
> I guess that env is required as I don't think we can use `${{ github.token }}` as default for env. vars. in the `action.yml`

Why would you need it for ENV vars?

<https://docs.github.com/en/actions/security-guides/automatic-token-authentication>

> The token is also available in the `github.token` context.
> 
> **Important:** An action can access the `GITHUB_TOKEN` through the `github.token` context even if the workflow does not explicitly pass the `GITHUB_TOKEN` to the action.

If you look at the links I provided to another action, they just retrieve the token from the context:

`action.yaml`:

```yaml
inputs:
  token:
    description: 'The GitHub authentication token'
    default: ${{ github.token }}
```

`main.ts`:
```ts
import * as github from '@actions/github'
// ...
const inputs = {
      token: core.getInput('token'),
      repository: core.getInput('repository'),
// ...
const octokit = github.getOctokit(inputs.token)
```

---

So you should be able to support getting the token the same way, without relying on ENV? You seem to use it pretty much the same way here:

https://github.com/micalevisk/last-issue-action/blob/a779b00cf9770e42a5ff23207d40c68a3f02a94f/src/fetch-last-issue-info.ts#L28

**Comment by micalevisk** [comment]
so it should be availalbe either as a input or as a env. var. I prefer the latter because it sounds better semantically speaking (_token as an input_ of the action doesn't feels good to me)

**Comment by polarathene** [comment]
<https://docs.github.com/en/actions/learn-github-actions/contexts#github-context>

> `github.token`:
> - A token to authenticate on behalf of the GitHub App installed on your repository.
> - This is functionally equivalent to the `GITHUB_TOKEN` secret.

**TL;DR:**

- `actions/checkout` (Official GH Action), and many community actions use `token` input in `action.yml` and **default to `github.token` context**.
- `GITHUB_TOKEN` according to docs is **never an ENV, only a secret**.
- Your `GITHUB_TOKEN` ENV would only ever work when explicitly providing the token? (_maybe this was not the case when you originally wrote the action_)

---

FWIW, your change says the `GITHUB_TOKEN` ENV isn't available to private repos, but are you sure it still is anywhere else...?

It's [not listed in the official docs default ENV](https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables), and the [token authentication docs page](https://docs.github.com/en/actions/security-guides/automatic-token-authentication) makes no mention of ENV, but clearly references it as a secret (_hence requiring me to set the ENV `GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}`_).

To further support why you should prefer the token via context, official GH actions such as [`actions/checkout` handle it this way](https://github.com/actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/action.yml#L12-L24) (_and many other actions I've come across likewise use a `token` input in `action.yml`_). `actions/checkouts` then [references the token here](https://github.com/actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/src/input-helper.ts#L110), and you can see no use or support for a `GITHUB_TOKEN` ENV.

I would think it better to be consistent (not only with other actions handling tokens, but in the UX).

<details>
<summary>More justification to prefer getting token via context over custom ENV (earlier response)</summary>

https://github.com/micalevisk/last-issue-action/blob/a779b00cf9770e42a5ff23207d40c68a3f02a94f/src/index.ts#L11-L14

https://github.com/micalevisk/last-issue-action/blob/a779b00cf9770e42a5ff23207d40c68a3f02a94f/action.yml#L4-L10

You would just add it as an optional input in `action.yml`, it could be named whatever you believe is more semantically correct, but it's common to see `token` in actions elsewhere. Call it `githubToken` there too if you like.

Then use the ENV as a fallback? (_or the other way, since the `default` for the input would probably always be populated anyway.._)

https://github.com/micalevisk/last-issue-action/blob/a779b00cf9770e42a5ff23207d40c68a3f02a94f/src/fetch-last-issue-info.ts#L28

```diff
- const octokitClient = makeOctokitClient(environment.githubToken); 
+ const octokitClient = makeOctokitClient(environment.githubToken || core.getInput('githubToken')); 
```

Personally it makes more sense to grab it from the `github.token` context, since other actions acquiring that in their `action.yml` are working fine without an explicit ENV being defined to pass a secret (_and the otherwise confusing 404 failure this issue is reporting, instead of expected 403 for auth failure_).

I haven't looked into it, but I think `GITHUB_TOKEN` may have been an ENV in the past? Or may have been how actions preferred to get the token until `github.token` as a default to the `action.yml` was a thing?

It doesn't sound as useful when the ENV is not implicit, and requires the user to set it with `${{ secrets.GITHUB_TOKEN }}`, whereas if actions are permitted to this token via `${{ github.token }}` in `action.yml` but not access to secrets implicitly, users are likely to feel more at ease security wise if the action itself gets the token via means it is authorized to use, and make that requirement explicitly defined in `action.yml`.

</details>

---

> I prefer the latter because it sounds better semantically speaking (token as an input of the action doesn't feels good to me)

It's what other actions do, but if you are concerned about an input name that is likely rarely required for a user to specify, then name it whatever you want (eg: `githubToken`). If your concern was with getting the default value from context `github.token`, then reference the docs in a comment above the `default` field?

**Comment by micalevisk** [comment]
feel free to open a PR to switch that. It will require a major bump tho.

tbh I don't see any real improvement on DX here but yeah. I believe in the past the `GITHUB_ENV` as a env. var was a thing on github actions out there

**Comment by polarathene** [comment]
> tbh I don't see any real improvement on DX here

- The user doesn't need to explicitly provide any token, it's **handled implicitly by the action** (_which was the original intent with the ENV_).
  - **Better DX from consistency:** If the ENV did work on public repos, but doesn't for private repos, while the `github.token` context is not affected? (_although I'm pretty certain public repos require the same treatment_)
  - **No surprises caused by a misleading `404` error** for the user to troubleshoot/investigate (_or they report and defer to you resolve_).
- **Using `github.token` context is terser:** Since the ENV is likely mandatory even for public repos.
- **Less security concerns raised:** Passing a secret to an action, that should already have authorized access to that secret via the `github` context object.
- **This issue itself wouldn't exist** and we'd both have saved time because the action _"just worked"_.

All perhaps minor DX, but still real improvements.

---

> feel free to open a PR to switch that.

Sure :+1: (**EDIT:** [Done](https://github.com/micalevisk/last-issue-action/pull/6))

> It will require a major bump tho.

No worries :)
